File: go/core/options.go
package core

import (
	"net/http"
	"time"
)

// ClientOptions configures the HTTP client shared across all service operations.
type ClientOptions struct {
	// APIKey for authentication. Required.
	APIKey string
	// BaseURL for API requests. Defaults to "https://runapi.ai".
	BaseURL string
	// Timeout for each HTTP request. Defaults to 15 minutes.
	Timeout time.Duration
	// MaxRetries is the maximum number of retry attempts. Defaults to 2.
	MaxRetries int
	// RetryBaseDelay is the base delay between retries. Defaults to 500ms.
	RetryBaseDelay time.Duration
	// RetryMaxDelay is the maximum delay between retries. Defaults to 5s.
	RetryMaxDelay time.Duration
	// HTTPClient is an optional custom *http.Client. When set, Timeout is ignored.
	HTTPClient *http.Client
	// UserAgent overrides the default User-Agent header.
	UserAgent string
	// Headers are extra HTTP headers sent with every request.
	Headers map[string]string
}

// RequestOptions are per-request overrides that take precedence over client-level defaults.
type RequestOptions struct {
	// Headers are additional HTTP headers merged with client-level headers.
	Headers map[string]string
	// Timeout overrides the client-level timeout for this request.
	Timeout time.Duration
	// MaxRetries overrides the client-level max retries for this request.
	MaxRetries *int
}

// PollingOptions controls async task polling behavior.
type PollingOptions struct {
	// PollInterval is the delay between poll requests. Defaults to 2s.
	PollInterval time.Duration
	// MaxWait is the maximum total wait time. Defaults to 15 minutes.
	MaxWait time.Duration
}

// CallConfig holds resolved request and polling options for a single API call.
// Service methods build one via ResolveCallOptions before issuing the request.
type CallConfig struct {
	// Request holds per-call HTTP overrides (timeout, retries, extra headers).
	Request RequestOptions
	// Polling controls async task polling intervals and maximum wait time.
	Polling PollingOptions
}

// CallOption is implemented by options that configure per-call behavior.
type CallOption interface {
	ApplyCall(*CallConfig)
}

// DefaultClientOptions returns ClientOptions with production defaults.
func DefaultClientOptions() ClientOptions {
	return ClientOptions{
		BaseURL:        DefaultBaseURL,
		Timeout:        DefaultTimeout,
		MaxRetries:     DefaultMaxRetries,
		RetryBaseDelay: DefaultRetryBaseDelay,
		RetryMaxDelay:  DefaultRetryMaxDelay,
	}
}

// DefaultPollingOptions returns PollingOptions with production defaults.
func DefaultPollingOptions() PollingOptions {
	return PollingOptions{PollInterval: DefaultPollInterval, MaxWait: DefaultMaxWait}
}

// ResolveCallOptions applies the given options to a CallConfig with default polling settings.
func ResolveCallOptions(options ...CallOption) CallConfig {
	cfg := CallConfig{Polling: DefaultPollingOptions()}
	for _, option := range options {
		if option != nil {
			option.ApplyCall(&cfg)
		}
	}
	return cfg
}

File: python/src/runapi/core/options.py
"""Client, request, and polling option dataclasses."""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional

from . import config, constants


@dataclass
class ClientOptions:
    """Configuration for an API client.

    ``http_client`` lets callers inject a custom transport (for tests or
    advanced use). When set, the SDK delegates all HTTP calls to it instead of
    building its own :class:`HttpClient`. It must expose
    ``request(method, path, body=None, options=None)``.
    """

    api_key: Optional[str] = None
    base_url: Optional[str] = None
    timeout: Optional[int] = None
    max_retries: Optional[int] = None
    retry_base_delay: Optional[float] = None
    retry_max_delay: Optional[float] = None
    http_client: Optional[Any] = None

    def __post_init__(self) -> None:
        if self.base_url is None:
            self.base_url = config.base_url
        if self.timeout is None:
            self.timeout = constants.HTTP_REQUEST_TIMEOUT
        if self.max_retries is None:
            self.max_retries = constants.MAX_RETRIES
        if self.retry_base_delay is None:
            self.retry_base_delay = constants.RETRY_BASE_DELAY
        if self.retry_max_delay is None:
            self.retry_max_delay = constants.RETRY_MAX_DELAY


@dataclass
class RequestOptions:
    """Per-request overrides of client-level defaults."""

    headers: Optional[Dict[str, str]] = None
    timeout: Optional[int] = None
    max_retries: Optional[int] = None


@dataclass
class PollingOptions:
    """Options for polling async task completion."""

    poll_interval: int = field(default=constants.POLLING_INTERVAL)
    max_wait: int = field(default=constants.POLLING_MAX_WAIT)
